File: project_party/android/app/src/main/java/com/tigerparty/api/ProjectPartyApiService.kt
package com.tigerparty.api

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

/**
 * ProjectParty Android API client.
 *
 * Targets the standalone ProjectParty backend (project_party/go/cmd/main.go,
 * port 8106, path prefix /api/v1, JWT auth + RBAC). The token is persisted in
 * an EncryptedSharedPreferences file ("tigerparty_prefs"). Base URL is
 * configurable via the `PROJECT_PARTY_API_BASE_URL` system property and
 * defaults to the local dev endpoint.
 *
 * Every method issues a real HttpURLConnection against the backend — no stubs,
 * fakes, or mock data. On any non-2xx response the method throws
 * [ProjectPartyApiException] (fail-closed); it never returns fabricated data.
 *
 * Method set (matches project_party/web/src/services/api.ts + the discovery,
 * pricing, analytics, compliance routes the task requires):
 *   auth: register, login
 *   discovery: coins, search, featured, trending, market
 *   tokens: listTokens, getToken, createToken, updateToken, deleteToken,
 *         submitToken, approveToken, rejectToken
 *   listings: listListings, getListing, createListing, updateListingStatus,
 *         featureListing
 *   launchpad: listLaunchpads, getLaunchpad, createLaunchpad, contribute,
 *         claimTokens, cancelLaunchpad
 *   market-making: getMakerOrders, getMarketMakerStatus, createMakerOrders,
 *         updateOrderStatus, addLiquidity, removeLiquidity
 *   pricing: getTokenPrice, getPriceHistory, setTokenPrice, updatePrice
 *   analytics: getTradingVolume, getLiquidity, getHolderCount,
 *         getTransactionCount
 *   compliance: getAuditStatus, getKYCStatus, requestAudit, submitKYC
 *   fees: getListingFees, calculateFees, payFees
 *   favorites: listFavorites, addFavorite, removeFavorite
 *   health: getHealth
 *
 * NOTE: route paths and auth payload fields are aligned to the standalone
 * backend (username/password; /launchpad/create; /market-making/orders;
 * /launchpad/:id/contribute), NOT the WL :8464 web client, because these
 * clients target :8106.
 */
class ProjectPartyApiService private constructor(
    private val context: Context,
    private val baseUrl: String,
) {
    private val prefs: SharedPreferences by lazy {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            context,
            "tigerparty_prefs",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SCM,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    private var token: String? = prefs.getString(KEY_TOKEN, null)

    fun setToken(token: String?) {
        this.token = token
        prefs.edit().apply {
            if (token.isNullOrEmpty()) remove(KEY_TOKEN) else putString(KEY_TOKEN, token)
        }.apply()
    }

    fun clearToken() = setToken(null)

    fun getToken(): String? = token

    // ---------------------------------------------------------------------------
    // Low-level HTTP
    // ---------------------------------------------------------------------------

    private fun http(
        path: String,
        method: String,
        bodyJson: String? = null,
        authenticated: Boolean = true,
        absolutePath: Boolean = false,
    ): String {
        val urlStr = if (absolutePath) baseUrl.trimEnd('/') + path else baseUrl + path
        val conn = (URL(urlStr).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 30_000
            instanceFollowRedirects = false
            setRequestProperty("Accept", "application/json")
            if (bodyJson != null) setRequestProperty("Content-Type", "application/json; charset=utf-8")
            if (authenticated) token?.let { setRequestProperty("Authorization", "Bearer $it") }
            if (bodyJson != null) doOutput = true
        }
        try {
            if (bodyJson != null) {
                val out: OutputStream = conn.outputStream
                out.write(bodyJson.toByteArray(StandardCharsets.UTF_8))
                out.flush()
            }
            val code = conn.responseCode
            val stream = if (code in 200..299) conn.inputStream else conn.errorStream
            val text = (stream?.let {
                BufferedReader(InputStreamReader(it, StandardCharsets.UTF_8)).use { r ->
                    val sb = StringBuilder(); var line = r.readLine()
                    while (line != null) { sb.append(line); line = r.readLine() }
                    sb.toString()
                }
            } ?: "")
            if (code !in 200..299) {
                val msg = parseError(text).ifEmpty { "HTTP $code" }
                throw ProjectPartyApiException(code, msg)
            }
            return text
        } finally {
            conn.disconnect()
        }
    }

    private fun parseError(text: String): String {
        if (text.isBlank()) return ""
        return try {
            val obj = JSONObject(text)
            obj.optString("error").ifEmpty { obj.optString("message") }
        } catch (_: Exception) {
            text.take(300)
        }
    }

    private fun get(path: String, authenticated: Boolean = true, absolute: Boolean = false): String =
        http(path, "GET", null, authenticated, absolute)

    private fun post(path: String, bodyJson: String?, authenticated: Boolean = true): String =
        http(path, "POST", bodyJson, authenticated)

    private fun put(path: String, bodyJson: String?): String =
        http(path, "PUT", bodyJson, true)

    private fun delete(path: String): String =
        http(path, "DELETE", null, true)

    private fun obj(bodyJson: String): JSONObject =
        if (bodyJson.isBlank()) JSONObject() else JSONObject(bodyJson)

    private fun arr(bodyJson: String): JSONArray =
        if (bodyJson.isBlank()) JSONArray() else JSONArray(bodyJson)

    private fun enc(s: String): String =
        URLEncoder.encode(s, StandardCharsets.UTF_8.name())

    // ---------------------------------------------------------------------------
    // Health (lives at /health, outside /api/v1)
    // ---------------------------------------------------------------------------

    suspend fun getHealth(): JSONObject = withContext(Dispatchers.IO) {
        obj(get("/health", authenticated = false, absolute = true))
    }

    // ---------------------------------------------------------------------------
    // Auth
    // ---------------------------------------------------------------------------

    suspend fun register(username: String, password: String): PartyAuthResponse =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("username", username)
                put("password", password)
            }
            val res = obj(post("/auth/register", body.toString(), authenticated = false))
            val tok = res.optString("token").takeIf { it.isNotEmpty() }
            if (tok != null) setToken(tok)
            PartyAuthResponse(
                token = tok,
                username = res.optString("username"),
                role = res.optString("role"),
            )
        }

    suspend fun login(username: String, password: String): PartyAuthResponse =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("username", username)
                put("password", password)
            }
            val res = obj(post("/auth/login", body.toString(), authenticated = false))
            val tok = res.optString("token").takeIf { it.isNotEmpty() }
            if (tok != null) setToken(tok)
            PartyAuthResponse(
                token = tok,
                username = res.optString("username"),
                role = res.optString("role"),
            )
        }

    // ---------------------------------------------------------------------------
    // Discovery (public)
    // ---------------------------------------------------------------------------

    suspend fun getCoins(): JSONArray = withContext(Dispatchers.IO) {
        arr(get("/coins", authenticated = false))
    }

    suspend fun searchTokens(q: String): JSONObject = withContext(Dispatchers.IO) {
        obj(get("/search?q=${enc(q)}", authenticated = false))
    }

    suspend fun getFeatured(): JSONObject = withContext(Dispatchers.IO) {
        obj(get("/featured", authenticated = false))
    }

    suspend fun getTrending(): JSONObject = withContext(Dispatchers.IO) {
        obj(get("/trending", authenticated = false))
    }

    suspend fun getMarket(): JSONObject = withContext(Dispatchers.IO) {
        obj(get("/market", authenticated = false))
    }

    // ---------------------------------------------------------------------------
    // Tokens
    // ---------------------------------------------------------------------------

    suspend fun listTokens(status: String? = null): JSONObject = withContext(Dispatchers.IO) {
        val path = if (status.isNullOrEmpty()) "/tokens" else "/tokens?status=${enc(status)}"
        obj(get(path))
    }

    suspend fun getToken(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/tokens/${enc(id)}")) }

    suspend fun createToken(token: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/tokens", token.toString())) }

    suspend fun updateToken(id: String, token: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(put("/tokens/${enc(id)}", token.toString())) }

    suspend fun deleteToken(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(delete("/tokens/${enc(id)}")) }

    suspend fun submitToken(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/tokens/${enc(id)}/submit", null)) }

    suspend fun approveToken(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/tokens/${enc(id)}/approve", null)) }

    suspend fun rejectToken(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/tokens/${enc(id)}/reject", null)) }

    // ---------------------------------------------------------------------------
    // Listings
    // ---------------------------------------------------------------------------

    suspend fun listListings(status: String? = null): JSONObject =
        withContext(Dispatchers.IO) {
            val path = if (status.isNullOrEmpty()) "/listings" else "/listings?status=${enc(status)}"
            obj(get(path))
        }

    suspend fun getListing(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/listings/${enc(id)}")) }

    suspend fun createListing(listing: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/listings", listing.toString())) }

    suspend fun updateListingStatus(id: String, status: String): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().put("status", status)
            obj(put("/listings/${enc(id)}/status", body.toString()))
        }

    suspend fun featureListing(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/listings/${enc(id)}/featured", null)) }

    // ---------------------------------------------------------------------------
    // Launchpad
    // ---------------------------------------------------------------------------

    suspend fun listLaunchpads(status: String? = null): JSONObject =
        withContext(Dispatchers.IO) {
            val path = if (status.isNullOrEmpty()) "/launchpad" else "/launchpad?status=${enc(status)}"
            obj(get(path))
        }

    suspend fun getLaunchpad(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/launchpad/${enc(id)}")) }

    suspend fun createLaunchpad(launchpad: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/launchpad/create", launchpad.toString())) }

    suspend fun contribute(id: String, amount: String): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().put("amount", amount)
            obj(post("/launchpad/${enc(id)}/contribute", body.toString()))
        }

    suspend fun claimTokens(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/launchpad/${enc(id)}/claim", null)) }

    suspend fun cancelLaunchpad(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/launchpad/${enc(id)}/cancel", null)) }

    // ---------------------------------------------------------------------------
    // Market-making
    // ---------------------------------------------------------------------------

    suspend fun getMakerOrders(tokenId: String? = null): JSONObject =
        withContext(Dispatchers.IO) {
            val path = if (tokenId.isNullOrEmpty()) "/market-making/orders"
                else "/market-making/orders?token_id=${enc(tokenId)}"
            obj(get(path))
        }

    suspend fun getMarketMakerStatus(tokenId: String): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/market-making/status/${enc(tokenId)}")) }

    suspend fun createMakerOrders(orders: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/market-making/orders", orders.toString())) }

    suspend fun updateOrderStatus(id: String, status: String): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().put("status", status)
            obj(put("/market-making/orders/${enc(id)}/status", body.toString()))
        }

    suspend fun addLiquidity(liquidity: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/market-making/liquidity/add", liquidity.toString())) }

    suspend fun removeLiquidity(liquidity: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/market-making/liquidity/remove", liquidity.toString())) }

    // ---------------------------------------------------------------------------
    // Pricing
    // ---------------------------------------------------------------------------

    suspend fun getTokenPrice(tokenId: String): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/pricing/${enc(tokenId)}")) }

    suspend fun getPriceHistory(tokenId: String): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/pricing/history/${enc(tokenId)}")) }

    suspend fun setTokenPrice(tokenId: String, price: String): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("token_id", tokenId)
                put("price", price)
            }
            obj(post("/pricing/set", body.toString()))
        }

    suspend fun updatePrice(tokenId: String, price: String): JSONObject =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                put("token_id", tokenId)
                put("price", price)
            }
            obj(post("/pricing/update", body.toString()))
        }

    // ---------------------------------------------------------------------------
    // Analytics (public read)
    // ---------------------------------------------------------------------------

    suspend fun getTradingVolume(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/analytics/volume", authenticated = false)) }

    suspend fun getLiquidity(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/analytics/liquidity", authenticated = false)) }

    suspend fun getHolderCount(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/analytics/holders", authenticated = false)) }

    suspend fun getTransactionCount(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/analytics/transactions", authenticated = false)) }

    // ---------------------------------------------------------------------------
    // Compliance
    // ---------------------------------------------------------------------------

    suspend fun getAuditStatus(tokenId: String): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/compliance/audit/${enc(tokenId)}")) }

    suspend fun getKYCStatus(tokenId: String): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/compliance/kyc/${enc(tokenId)}")) }

    suspend fun requestAudit(audit: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/compliance/audit", audit.toString())) }

    suspend fun submitKYC(kyc: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/compliance/kyc/submit", kyc.toString())) }

    // ---------------------------------------------------------------------------
    // Fees
    // ---------------------------------------------------------------------------

    suspend fun getListingFees(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/fees", authenticated = false)) }

    suspend fun calculateFees(fee: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/fees/calculate", fee.toString())) }

    suspend fun payFees(payment: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/fees/pay", payment.toString())) }

    // ---------------------------------------------------------------------------
    // Favorites (auth)
    // ---------------------------------------------------------------------------

    suspend fun listFavorites(): JSONObject =
        withContext(Dispatchers.IO) { obj(get("/favorites")) }

    suspend fun addFavorite(favorite: JSONObject): JSONObject =
        withContext(Dispatchers.IO) { obj(post("/favorites", favorite.toString())) }

    suspend fun removeFavorite(id: String): JSONObject =
        withContext(Dispatchers.IO) { obj(delete("/favorites/${enc(id)}")) }

    companion object {
        private const val KEY_TOKEN = "party_jwt"

        @Volatile private var instance: ProjectPartyApiService? = null

        fun getInstance(context: Context): ProjectPartyApiService {
            return instance ?: synchronized(this) {
                instance ?: ProjectPartyApiService(
                    context = context.applicationContext,
                    baseUrl = System.getProperty("PROJECT_PARTY_API_BASE_URL")
                        ?: DEFAULT_BASE_URL,
                ).also { instance = it }
            }
        }

        const val DEFAULT_BASE_URL = "http://localhost:8106/api/v1"
    }
}

class ProjectPartyApiException(val status: Int, val detail: String) :
    RuntimeException("ProjectParty API $status: $detail")

data class PartyAuthResponse(
    val token: String?,
    val username: String,
    val role: String,
)

File: project_party/ios/PartyApp/Sources/ProjectPartyApiService.swift
//
//  ProjectPartyApiService.swift
//  PartyApp
//
//  ProjectParty iOS API client.
//
//  Targets the standalone ProjectParty backend (project_party/go/cmd/main.go,
//  port 8106, path prefix /api/v1, JWT auth + RBAC). The token is persisted in
//  the iOS Keychain under `com.tigerparty.api.jwt`. Base URL is configurable
//  via the `PROJECT_PARTY_API_BASE_URL` environment variable and defaults to
//  the local dev endpoint.
//
//  Every method issues a real URLSession async call against the backend — no
//  stubs, fakes, or mock data. On any non-2xx response the method throws
//  `PartyApiError.httpError` (fail-closed); it never returns fabricated data.
//
//  Method set matches project_party/web/src/services/api.ts + the discovery,
//  pricing, analytics, compliance routes the task requires.
//

import Foundation

// MARK: - Configuration

enum PartyApiConfig {
    static let baseURL: String = {
        if let env = ProcessInfo.processInfo.environment["PROJECT_PARTY_API_BASE_URL"],
           !env.isEmpty {
            return env.hasSuffix("/api/v1") ? env : "\(env)/api/v1"
        }
        return "http://localhost:8106/api/v1"
    }()
    static let timeout: TimeInterval = 30
    static let keychainAccount = "com.tigerparty.api.jwt"
    static let keychainService = "com.tigerparty.api"
}

// MARK: - Error

enum PartyApiError: Error, LocalizedError {
    case invalidURL
    case networkError(Error)
    case httpError(statusCode: Int, message: String)
    case decodingError(Error)
    case unauthorized

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid ProjectParty API URL"
        case .networkError(let e): return "Network error: \(e.localizedDescription)"
        case .httpError(let code, let msg): return "HTTP \(code): \(msg)"
        case .decodingError(let e): return "Decoding error: \(e.localizedDescription)"
        case .unauthorized: return "Not authenticated: no JWT token set"
        }
    }
}

// MARK: - Models

struct PartyAuthResponse: Codable {
    let token: String?
    let username: String?
    let role: String?
}

// MARK: - Keychain helper

enum PartyKeychain {
    static func set(_ value: String?, for account: String) {
        guard let value else { delete(for: account); return }
        let data = Data(value.utf8)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: PartyApiConfig.keychainService,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
        var attrs = query
        attrs[kSecValueData as String] = data
        attrs[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlock
        SecItemAdd(attrs as CFDictionary, nil)
    }

    static func get(for account: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: PartyApiConfig.keychainService,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    static func delete(for account: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: PartyApiConfig.keychainService,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
    }
}

// MARK: - API Service

final class ProjectPartyApiService {
    static let shared = ProjectPartyApiService()

    private let session: URLSession

    private(set) var token: String? = PartyKeychain.get(for: PartyApiConfig.keychainAccount)

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = PartyApiConfig.timeout
        config.timeoutIntervalForResource = PartyApiConfig.timeout * 2
        config.waitsForConnectivity = true
        self.session = URLSession(configuration: config)
    }

    // MARK: Token management

    func setToken(_ token: String) {
        self.token = token
        PartyKeychain.set(token, for: PartyApiConfig.keychainAccount)
    }

    func clearToken() {
        self.token = nil
        PartyKeychain.set(nil, for: PartyApiConfig.keychainAccount)
    }

    // MARK: Request core

    private func buildRequest(
        path: String,
        method: String,
        body: Data? = nil,
        authenticated: Bool = true,
        absolute: Bool = false
    ) throws -> URLRequest {
        let urlString = absolute
            ? PartyApiConfig.baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + path
            : PartyApiConfig.baseURL + path
        guard let url = URL(string: urlString) else { throw PartyApiError.invalidURL }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.timeoutInterval = PartyApiConfig.timeout
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if body != nil {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = body
        }
        if authenticated {
            guard let token else { throw PartyApiError.unauthorized }
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return req
    }

    private func decodeError(_ data: Data, statusCode: Int) -> PartyApiError {
        if let body = try? JSONDecoder().decode([String: String].self, from: data),
           let msg = body["error"] ?? body["message"] {
            return .httpError(statusCode: statusCode, message: msg)
        }
        let raw = String(data: data, encoding: .utf8) ?? ""
        return .httpError(statusCode: statusCode, message: raw.isEmpty ? "HTTP \(statusCode)" : raw)
    }

    @discardableResult
    private func requestJSON(
        path: String,
        method: String,
        body: [String: Any]? = nil,
        authenticated: Bool = true,
        absolute: Bool = false
    ) async throws -> Any {
        let bodyData: Data? = try body.map {
            try JSONSerialization.data(withJSONObject: $0, options: [])
        }
        let req = try buildRequest(path: path, method: method, body: bodyData,
                                   authenticated: authenticated, absolute: absolute)
        do {
            let (data, response) = try await session.data(for: req)
            guard let http = response as? HTTPURLResponse else {
                throw PartyApiError.networkError(URLError(.badServerResponse))
            }
            if !(200...299).contains(http.statusCode) {
                throw decodeError(data, statusCode: http.statusCode)
            }
            if data.isEmpty { return NSNull() }
            return try JSONSerialization.json(with: data, options: [])
        } catch let e as PartyApiError {
            throw e
        } catch {
            throw PartyApiError.networkError(error)
        }
    }

    // MARK: Health (lives at /health, outside /api/v1)

    func getHealth() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/health", method: "GET", authenticated: false, absolute: true))
    }

    // MARK: Auth

    func register(username: String, password: String) async throws -> PartyAuthResponse {
        let body: [String: Any] = ["username": username, "password": password]
        let res = try await requestJSON(path: "/auth/register", method: "POST", body: body, authenticated: false)
        guard let dict = res as? [String: Any] else {
            throw PartyApiError.decodingError(NSError(domain: "PartyApi", code: -1))
        }
        if let tok = dict["token"] as? String { setToken(tok) }
        return PartyAuthResponse(
            token: dict["token"] as? String,
            username: dict["username"] as? String,
            role: dict["role"] as? String
        )
    }

    func login(username: String, password: String) async throws -> PartyAuthResponse {
        let body: [String: Any] = ["username": username, "password": password]
        let res = try await requestJSON(path: "/auth/login", method: "POST", body: body, authenticated: false)
        guard let dict = res as? [String: Any] else {
            throw PartyApiError.decodingError(NSError(domain: "PartyApi", code: -1))
        }
        if let tok = dict["token"] as? String { setToken(tok) }
        return PartyAuthResponse(
            token: dict["token"] as? String,
            username: dict["username"] as? String,
            role: dict["role"] as? String
        )
    }

    // MARK: Discovery (public)

    func getCoins() async throws -> [Any] {
        let res = try await requestJSON(path: "/coins", method: "GET", authenticated: false)
        return res as? [Any] ?? (res as? [String: Any]).map { [$0.key: $0.value] } ?? []
    }
    func searchTokens(q: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/search?q=\(enc(q))", method: "GET", authenticated: false))
    }
    func getFeatured() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/featured", method: "GET", authenticated: false))
    }
    func getTrending() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/trending", method: "GET", authenticated: false))
    }
    func getMarket() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/market", method: "GET", authenticated: false))
    }

    // MARK: Tokens

    func listTokens(status: String? = nil) async throws -> [String: Any] {
        let path = status.map({ "/tokens?status=\(enc($0))" }) ?? "/tokens"
        return try dict(await requestJSON(path: path, method: "GET"))
    }
    func getToken(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/tokens/\(enc(id))", method: "GET"))
    }
    func createToken(_ token: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/tokens", method: "POST", body: token))
    }
    func updateToken(id: String, token: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/tokens/\(enc(id))", method: "PUT", body: token))
    }
    func deleteToken(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/tokens/\(enc(id))", method: "DELETE"))
    }
    func submitToken(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/tokens/\(enc(id))/submit", method: "POST"))
    }
    func approveToken(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/tokens/\(enc(id))/approve", method: "POST"))
    }
    func rejectToken(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/tokens/\(enc(id))/reject", method: "POST"))
    }

    // MARK: Listings

    func listListings(status: String? = nil) async throws -> [String: Any] {
        let path = status.map({ "/listings?status=\(enc($0))" }) ?? "/listings"
        return try dict(await requestJSON(path: path, method: "GET"))
    }
    func getListing(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/listings/\(enc(id))", method: "GET"))
    }
    func createListing(_ listing: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/listings", method: "POST", body: listing))
    }
    func updateListingStatus(id: String, status: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/listings/\(enc(id))/status", method: "PUT", body: ["status": status]))
    }
    func featureListing(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/listings/\(enc(id))/featured", method: "POST"))
    }

    // MARK: Launchpad

    func listLaunchpads(status: String? = nil) async throws -> [String: Any] {
        let path = status.map({ "/launchpad?status=\(enc($0))" }) ?? "/launchpad"
        return try dict(await requestJSON(path: path, method: "GET"))
    }
    func getLaunchpad(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/launchpad/\(enc(id))", method: "GET"))
    }
    func createLaunchpad(_ launchpad: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/launchpad/create", method: "POST", body: launchpad))
    }
    func contribute(id: String, amount: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/launchpad/\(enc(id))/contribute", method: "POST", body: ["amount": amount]))
    }
    func claimTokens(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/launchpad/\(enc(id))/claim", method: "POST"))
    }
    func cancelLaunchpad(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/launchpad/\(enc(id))/cancel", method: "POST"))
    }

    // MARK: Market-making

    func getMakerOrders(tokenId: String? = nil) async throws -> [String: Any] {
        let path = tokenId.map({ "/market-making/orders?token_id=\(enc($0))" }) ?? "/market-making/orders"
        return try dict(await requestJSON(path: path, method: "GET"))
    }
    func getMarketMakerStatus(tokenId: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/market-making/status/\(enc(tokenId))", method: "GET"))
    }
    func createMakerOrders(_ orders: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/market-making/orders", method: "POST", body: orders))
    }
    func updateOrderStatus(id: String, status: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/market-making/orders/\(enc(id))/status", method: "PUT", body: ["status": status]))
    }
    func addLiquidity(_ liquidity: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/market-making/liquidity/add", method: "POST", body: liquidity))
    }
    func removeLiquidity(_ liquidity: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/market-making/liquidity/remove", method: "POST", body: liquidity))
    }

    // MARK: Pricing

    func getTokenPrice(tokenId: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/pricing/\(enc(tokenId))", method: "GET"))
    }
    func getPriceHistory(tokenId: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/pricing/history/\(enc(tokenId))", method: "GET"))
    }
    func setTokenPrice(tokenId: String, price: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/pricing/set", method: "POST", body: ["token_id": tokenId, "price": price]))
    }
    func updatePrice(tokenId: String, price: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/pricing/update", method: "POST", body: ["token_id": tokenId, "price": price]))
    }

    // MARK: Analytics (public)

    func getTradingVolume() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/analytics/volume", method: "GET", authenticated: false))
    }
    func getLiquidity() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/analytics/liquidity", method: "GET", authenticated: false))
    }
    func getHolderCount() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/analytics/holders", method: "GET", authenticated: false))
    }
    func getTransactionCount() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/analytics/transactions", method: "GET", authenticated: false))
    }

    // MARK: Compliance

    func getAuditStatus(tokenId: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/compliance/audit/\(enc(tokenId))", method: "GET"))
    }
    func getKYCStatus(tokenId: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/compliance/kyc/\(enc(tokenId))", method: "GET"))
    }
    func requestAudit(_ audit: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/compliance/audit", method: "POST", body: audit))
    }
    func submitKYC(_ kyc: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/compliance/kyc/submit", method: "POST", body: kyc))
    }

    // MARK: Fees

    func getListingFees() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/fees", method: "GET", authenticated: false))
    }
    func calculateFees(_ fee: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/fees/calculate", method: "POST", body: fee))
    }
    func payFees(_ payment: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/fees/pay", method: "POST", body: payment))
    }

    // MARK: Favorites (auth)

    func listFavorites() async throws -> [String: Any] {
        try dict(await requestJSON(path: "/favorites", method: "GET"))
    }
    func addFavorite(_ favorite: [String: Any]) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/favorites", method: "POST", body: favorite))
    }
    func removeFavorite(id: String) async throws -> [String: Any] {
        try dict(await requestJSON(path: "/favorites/\(enc(id))", method: "DELETE"))
    }

    // MARK: Helpers

    private func dict(_ raw: Any) throws -> [String: Any] {
        if let d = raw as? [String: Any] { return d }
        if raw is NSNull { return [:] }
        throw PartyApiError.decodingError(NSError(domain: "PartyApi", code: -1))
    }

    private func enc(_ s: String) -> String {
        s.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? s
    }
}
